daphne/src/messages.rs

```diff
@@ -809,6 +822,266 @@ impl Decode for HpkeCiphertext {
+        })
+    }
+}
+
+// The variable-length part of a QueryConfig in draft-wang-ppm-dap-taskprov
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
+pub enum QueryConfigVar {
+    TimeInterval,
+    FixedSize { max_batch_size: u32 },
+}
+
+// There is no Encode or Decode for QueryConfigVar as we have to split the query type and
+// the variable length stuff in the message format, so we do all of the work in QueryConfig's
+// Encode and Decode.
+
+// The query configuration of draft-wang-ppm-dap-taskprov
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
+pub struct QueryConfig {
+    pub time_precision: Duration,
+    pub max_batch_query_count: u16,
+    pub min_batch_size: u32,
+    pub var: QueryConfigVar,
+}
+
+impl Encode for QueryConfig {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        match &self.var {
+            QueryConfigVar::TimeInterval { .. } => {
+                QUERY_TYPE_TIME_INTERVAL.encode(bytes);
+            }
+            QueryConfigVar::FixedSize { .. } => {
+                QUERY_TYPE_FIXED_SIZE.encode(bytes);
+            }
+        }
+        self.time_precision.encode(bytes);
+        self.max_batch_query_count.encode(bytes);
+        self.min_batch_size.encode(bytes);
+        match &self.var {
+            QueryConfigVar::TimeInterval { .. } => {
+                // Nothing to do.
+            }
+            QueryConfigVar::FixedSize { max_batch_size } => {
+                max_batch_size.encode(bytes);
+            }
+        }
+    }
+}
+
+impl Decode for QueryConfig {
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        let query_type = u8::decode(bytes)?;
+        let time_precision = Duration::decode(bytes)?;
+        let max_batch_query_count = u16::decode(bytes)?;
+        let min_batch_size = u32::decode(bytes)?;
+        let var = match query_type {
+            QUERY_TYPE_TIME_INTERVAL => Ok(QueryConfigVar::TimeInterval),
+            QUERY_TYPE_FIXED_SIZE => Ok(QueryConfigVar::FixedSize {
+                max_batch_size: u32::decode(bytes)?,
+            }),
+            _ => Err(CodecError::UnexpectedValue),
+        }?;
+        Ok(Self {
+            time_precision,
+            max_batch_query_count,
+            min_batch_size,
+            var,
+        })
+    }
+}
+
+// The Task configuration of draft-wang-ppm-dap-taskprov
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
+pub struct TaskConfig {
+    pub task_info: Vec<u8>,
+    pub aggregator_endpoints: Vec<Vec<u8>>,
+    pub query_config: QueryConfig,
+    pub task_expiration: Time,
+    pub vdaf_config: VdafConfig,
+}
+
+impl Encode for TaskConfig {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        encode_u8_items(bytes, &(), &self.task_info);
+        let url_bytes: Vec<UrlBytes> = self
+            .aggregator_endpoints
+            .iter()
+            .map(|v| UrlBytes { bytes: v.clone() })
+            .collect();
```

Comment:
We wouldn't need to construct this if we used `Vec<UrlBytes>` as the type of `aggreator_endpoints`. This would be slightlier easier to read, and would avoid the clones.